File: src/alt_splice/detect_altprime.py
def detect_altprime(genes, idx_alt):
# function [idx_alt_5prime,exon_alt_5prime, idx_alt_3prime,exon_alt_3prime] ...
#    = detect_altprime(genes, idx_alt);
#
# detect the alternative 5 and 3 prime ends of the intron. Note that 5 prime refers to the left
# and 3 prime to the right for a positive strand

MIN_OVERLAP = 11 # two exons that are alternative should overlap by at least the length

idx_alt_5prime = []
idx_alt_3prime = []
exon_alt_5prime = {}
exon_alt_3prime = {}

for ix in idx_alt:
    if idx % 50 == 0:
        print '.',
    num_exons = genes[ix].splicegraph.get_len()
    vertices = genes[ix].splicegraph.vertices
    edges = genes[ix].splicegraph.edges
    strand = genes[ix].strand
    # Find alternative sites on the right of the intron,
    # same site on the left of the intron.
    for exon_idx in range(num_exons - 2):
        rightsites = []
        rightidx = []
        nr_exons = sp.sum(edges[exon_idx, exon_idx + 1 : num_exons])
        if nr_exons >= 2:
            which_exons = sp.where(edges[exon_idx, exon_idx + 1 : num_exons])[0] + exon_idx
            exons = vertices[:, which_exons]
            for i in range(nr_exons - 1):
                for j in range(i + 1, nr_exons):
                    # check that the left splice site of the exons are different
                    # make sure exons overlap - either:
                    # - left splice site of exon(i) is in exon(j)
                    # - left splice site of exon(j) is in exon(i)
                    # note that the 'overlap' relationship is not transitive
                    if ((exons[0, i] != exons[0, j]) and 
                      (((exons[0, i] >= exons[0, j]) and (exons[0, i] <= exons[1, j])) or ((exons[0, j] >= exons[0, i]) and (exons[0, j] <= exons[1, i]))) and
                      (min(exons[1, i], exons[1, j]) - max(exons[0, i], exons[0, j]) >= MIN_OVERLAP)):

                        assert(not ((exons[0, i] == exons[0, j]) and (exons[1, i] == exons[1, j])));
                        assert(exons[0, i] != exons[0, j])

                        ### add new events to the list
                        if not exons[0, i] in rightsites:
                            rightsites.append(exons[0, i])
                            rightidx.append(which_exons[i])
                        if not exons[0, j] in rightsites:
                            rightsites.append(exons[0, j])
                            rightidx.append(which_exons[j])
       
        # construct the output
        if len(rightsites) >= 2:
            if strand == '+':
                exon_alt_3prime.append({'fiveprimesite':exon_idx, 'threeprimesites':rightidx})
                idx_alt_3prime.append(ix)
            if strand == '-':
                exon_alt_5prime.append({'threeprimesite':exon_idx, 'fiveprimesites':rightidx})
                idx_alt_5prime.append(ix)
    
    # Find alternative sites on the left of the intron,
    # same site on the right of the intron.
    for exon_idx in range(2, num_exons):
        nr_exons = sp.sum(edges[:exon_idx-1, exon_idx])
        leftsites = []
        leftidx = []
        if nr_exons >= 2:
            which_exons = sp.where(edges[:exon_idx-1, exon_idx])[0]
            exons = vertices[:, which_exons]
            for i in range(nr_exons - 1):
                for j in range(i + 1, nr_exons):
                    # check that the 5prime sites are different
                    # make sure exons overlap - either:
                    # - right splice site of exon(i) is in exon(j)
                    # - right splice site of exon(j) is in exon(i)
                    # note that the 'overlap' relationship is not transitive
                    if ((exons[1, i] != exons[1, j]) and
                      (((exons[1, i] <= exons[1, j]) and (exons[1, i] >= exons[0,j])) or ((exons[1, j] <= exons[1, i]) and (exons[1, j] >= exons[0, i]))) and
                      (min(exons[1, i], exons[1, j]) - max(exons[1, i], exons[0, j]) >= MIN_OVERLAP)):

                        assert(not((exons[0, i] == exons[0, j]) and (exons[1, i] == exons[1, j])))
                        assert(exons[1, i] != exons[1, j])
                    
                        # add new events to the list
                        if not exons[1, i] in leftsites:
                            leftsites.append(exons[1, i])
                            leftidx.append(which_exons[i])
                        if not exons[1, j] in leftsites:
                            leftsites.append(exons[1, j])
                            leftidx.append(which_exons[j])

        # construct the output
        if len(leftsites) >= 2:
            if strand=='+'
                exon_alt_5prime.append({'threeprimesite':exon_idx, 'fiveprimesites':leftidx})
                idx_alt_5prime.append(ix) 
            if strand =='-'
                exon_alt_3prime.append({'fiveprimesite':exon_idx, 'threeprimesites':leftidx})
                idx_alt_3prime.append(ix)

    print '\nNumber of alternative 5 prime sites:\t\t\t\t%d' % len(idx_alt_5prime)
    print 'Number of alternative 3 prime sites:\t\t\t\t%d' % len(idx_alt_3prime)

    return (idx_alt_5prime, exon_alt_5prime, idx_alt_3prime, exon_alt_3prime)

File: src/alt_splice/detect_altprime.m
function [idx_alt_5prime,exon_alt_5prime, idx_alt_3prime,exon_alt_3prime] ...
    = detect_altprime(genes, idx_alt);
% function [idx_alt_5prime,exon_alt_5prime, idx_alt_3prime,exon_alt_3prime] ...
%    = detect_altprime(genes, idx_alt);
%
% detect the alternative 5 and 3 prime ends of the intron. Note that 5 prime refers to the left
% and 3 prime to the right for a positive strand

MIN_OVERLAP = 11; % two exons that are alternative should overlap by at least the BLAT seed

idx_alt_5prime = [];
idx_alt_3prime = [];
exon_alt_5prime = {};
exon_alt_3prime = {};


for ix=idx_alt
  if (mod(ix,50)==0)
    fprintf(1,'.');
  end
  num_exons = size(genes(ix).splicegraph{1},2) ;
  vertices = genes(ix).splicegraph{1} ;
  edges = genes(ix).splicegraph{2} ;
  if isfield(genes(ix),'strands')
    strand = genes(ix).strands(1);
  else
    strand = genes(ix).strand;
  end
  % Find alternative sites on the right of the intron,
  % same site on the left of the intron.
  % Store the splice site at the left of the intron in 'starts'.
  starts=[] ;
  for exon_idx = 1:num_exons-2
    %if sum(find(starts==vertices(2,exon_idx)))==0,
    %  starts=[starts, vertices(2,exon_idx)] ;
      nr_exons=sum(edges(exon_idx,exon_idx+1:num_exons)) ;
      rightsites=[] ;
      rightidx=[] ;
      if nr_exons>=2,
	which_exons = find([zeros(1,exon_idx),edges(exon_idx,exon_idx+1:num_exons)]);
        exons=[vertices(:, which_exons)] ;
        for i=1:nr_exons-1
          for j=i+1:nr_exons
	    % check that the left splice site of the exons are different
	    % make sure exons overlap - either:
	    % - left splice site of exon(i) is in exon(j)
	    % - left splice site of exon(j) is in exon(i)
	    % note that the 'overlap' relationship is not transitive
	    if ( (exons(1,i) ~= exons(1,j)) && ...
	    	 (( (exons(1,i) >= exons(1,j)) && (exons(1,i) <= exons(2,j)) ) ...
	    	  || ( (exons(1,j) >= exons(1,i)) && (exons(1,j) <= exons(2,i)) ) ) && ...
	    	 (min(exons(2,i),exons(2,j)) - max(exons(1,i),exons(1,j)) + 1 >= MIN_OVERLAP ) )
	      assert(~((exons(1,i) == exons(1,j)) && (exons(2,i) == exons(2,j))));
	      assert(exons(1,i) ~= exons(1,j));
	    % check that the right (exon) sites are different but left splice is same.
	    %if ( (exons(2,i) ~= exons(2,j)) && (exons(1,i) == exons(1,j)) )
	      % add new events to the list
	      if sum(find(rightsites(1:end)==exons(1,i)))==0
                rightsites=[rightsites, exons(1,i)] ;
                rightidx=[rightidx, which_exons(i)] ;
              end
              if sum(find(rightsites(1:end)==exons(1,j)))==0
                rightsites=[rightsites, exons(1,j)] ;
                rightidx=[rightidx, which_exons(j)] ;
              end
            end  
          end
        end     
      end
     
      % construct the output
      if length(rightsites)>=2,
        if strand=='+',
          exon_alt_3prime(end+1).fiveprimesite=exon_idx ;
          exon_alt_3prime(end).threeprimesites=rightidx ;
          idx_alt_3prime = [idx_alt_3prime, ix] ;
        end
        if strand=='-',
          exon_alt_5prime(end+1).threeprimesite=exon_idx ;
          exon_alt_5prime(end).fiveprimesites=rightidx ;
          idx_alt_5prime = [idx_alt_5prime, ix] ;
        end
      end  % construct output
    %end
  end % for exon_idx

  
  % Find alternative sites on the left of the intron,
  % same site on the right of the intron.
  % Store the splice site at the right of the intron in 'starts'.
  starts=[] ;
  for exon_idx = 3:num_exons
    %if sum(find(starts==vertices(1,exon_idx)))==0,
    %  starts=[starts, vertices(1,exon_idx)] ;
      nr_exons=sum(edges(1:exon_idx-1,exon_idx)) ;
      leftsites=[] ;
      leftidx=[] ;
      if nr_exons>=2,
	    which_exons = find([edges(1:exon_idx-1,exon_idx)', zeros(1,num_exons-exon_idx+1)]);
        exons=[vertices(:, which_exons)] ;
        for i=1:nr_exons-1
          for j=i+1:nr_exons
	    % check that the 5prime sites are different
	    % make sure exons overlap - either:
	    % - right splice site of exon(i) is in exon(j)
	    % - right splice site of exon(j) is in exon(i)
	    % note that the 'overlap' relationship is not transitive
	    if ((exons(2,i) ~= exons(2,j)) &&...
	    	( ( (exons(2,i) <= exons(2,j)) && (exons(2,i) >= exons(1,j)) )...
	    	|| ( (exons(2,j) <= exons(2,i)) && (exons(2,j) >= exons(1,i)) ) ) && ...
	    	(min(exons(2,i),exons(2,j)) - max(exons(1,i),exons(1,j)) + 1 >= MIN_OVERLAP ) )
	      assert(~((exons(1,i) == exons(1,j)) && (exons(2,i) == exons(2,j))));
	      assert(exons(2,i) ~= exons(2,j));
	      
	    % check that the right (exon) sites are different but left splice is same.
	    %if ( (exons(2,i) ~= exons(2,j)) && (exons(1,i) == exons(1,j)) )
	      % add new events to the list
              if sum(find(leftsites(1:end)==exons(2,i)))==0
                leftsites=[leftsites, exons(2,i)] ;
                leftidx=[leftidx, which_exons(i)] ;
              end
              if sum(find(leftsites(1:end)==exons(2,j)))==0
                leftsites=[leftsites, exons(2,j)] ;
                leftidx=[leftidx, which_exons(j)] ;
              end
            end
          end
        end     
      end

      % construct the output
      if length(leftsites)>=2,
        if strand=='+'
          exon_alt_5prime(end+1).threeprimesite=exon_idx ;
          exon_alt_5prime(end).fiveprimesites=leftidx ;
          idx_alt_5prime = [idx_alt_5prime, ix] ;
        end
        if strand =='-'
          exon_alt_3prime(end+1).fiveprimesite=exon_idx ;
          exon_alt_3prime(end).threeprimesites=leftidx ;
          idx_alt_3prime = [idx_alt_3prime, ix] ;
        end
      end % construct output
    %end
  end % for exon_idx
  


end





fprintf(1,'\n\nNumber of alternative 5 prime sites:\t\t\t\t%d\n',...
	length(idx_alt_5prime));

fprintf(1,'Number of alternative 3 prime sites:\t\t\t\t%d\n',...
	length(idx_alt_3prime));



